lib/charms/prometheus_k8s/v0/prometheus_remote_write.py

```diff
@@ -1,93 +1,701 @@
+                via `relation_name` argument does not have the `prometheus_scrape` relation
+                interface.
+            RelationRoleMismatchError: If the relation with the same name as provided
+                via `relation_name` argument does not have the `RelationRole.requires`
+                role.
         """
+        _validate_relation_by_interface_and_direction(
+            charm, relation_name, RELATION_INTERFACE_NAME, RelationRole.requires
+        )
+
+        try:
+            alert_rules_path = _resolve_dir_against_charm_path(charm, alert_rules_path)
+        except InvalidAlertRuleFolderPathError as e:
+            logger.debug(
+                "Invalid Prometheus alert rules folder at %s: %s",
+                e.alert_rules_absolute_path,
+                e.message,
+            )
+
         super().__init__(charm, relation_name)
+        self._charm = charm
         self._relation_name = relation_name
+        self._alert_rules_path = alert_rules_path
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_joined,
+            self._handle_endpoints_changed,
+        )
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_changed,
+            self._handle_endpoints_changed,
+        )
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_departed,
+            self._handle_endpoints_changed,
+        )
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_broken,
+            self._handle_endpoints_changed,
+        )
+
+        self.framework.observe(
+            self._charm.on[self._relation_name].relation_joined,
+            self._set_alerts_on_relation_changed,
+        )
+        self.framework.observe(
+            self._charm.on.upgrade_charm,
+            self._set_alerts_to_all_relation,
+        )
+
+    def _handle_endpoints_changed(self, event: RelationEvent):
+        self.on.endpoints_changed.emit(relation_id=event.relation.id)
+
+    def _set_alerts_on_relation_changed(self, event: RelationEvent):
+        self._set_alerts_to_relation(event.relation)
+
+    def _set_alerts_to_all_relation(self, _):
+        for relation in self.model.relations[self._relation_name]:
+            self._set_alerts_to_relation(relation)
+
+    def _set_alerts_to_relation(self, relation: Relation):
+        if alert_groups := self._labeled_alert_groups:
+            relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": alert_groups})
 
     @property
-    def endpoints(self) -> Generator[str, None, None]:
-        """A list of remote write endpoints.
+    def endpoints(self) -> List[Dict[str, str]]:
+        """A config object ready to be dropped in to a prometheus config file.
 
         Returns:
-            A list of remote write endpoints.
+            A list of remote_write configs.
```

Comment:
A little confusing, since it's used as `.endpoints` elsewhere. Are they configs or endpoints?